File: app/src/main/java/cheema/hardeep/sahibdeep/studentify/activities/TasksActivity.java
package cheema.hardeep.sahibdeep.studentify.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import cheema.hardeep.sahibdeep.studentify.R;
import cheema.hardeep.sahibdeep.studentify.adapters.TaskAdapter;
import cheema.hardeep.sahibdeep.studentify.database.StudentifyDatabaseProvider;
import cheema.hardeep.sahibdeep.studentify.interfaces.ViewRefreshInterface;
import cheema.hardeep.sahibdeep.studentify.models.tables.Task;
import cheema.hardeep.sahibdeep.studentify.models.tables.TaskType;

import static cheema.hardeep.sahibdeep.studentify.activities.TasksDetailsActivity.NEGATIVE_TASK_ID;

public class TasksActivity extends AppCompatActivity implements ViewRefreshInterface {

    private static final String KEY_CLASS_ID = "key-class-id";

    public static Intent createIntent(Context context, int classId) {
        Intent intent = new Intent(context, TasksActivity.class);
        intent.putExtra(KEY_CLASS_ID, classId);
        return intent;
    }

    @BindView(R.id.addHomeworkButton)
    ImageView addHomeworkButton;

    @BindView(R.id.addTestButton)
    ImageView addTestButton;

    @BindView(R.id.homeworkRecyclerView)
    RecyclerView homeworkRecyclerView;

    @BindView(R.id.noTest)
    TextView noTest;

    @BindView(R.id.noHomework)
    TextView noHomework;

    @BindView(R.id.testRecyclerView)
    RecyclerView testRecyclerView;

    @BindView(R.id.doneButton)
    Button doneButton;

    TaskAdapter homeworkAdapter;
    TaskAdapter testAdapter;

    int classId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_tasks);
        getSupportActionBar().hide();
        ButterKnife.bind(this);

        classId = getIntent().getIntExtra(KEY_CLASS_ID, -1);
        setupHomeworkRV();
        setupTestRV();

        addHomeworkButton.setOnClickListener(v ->
                startActivity(TasksDetailsActivity.createIntent(v.getContext(), true, classId, NEGATIVE_TASK_ID)));
        addTestButton.setOnClickListener(v ->
                startActivity(TasksDetailsActivity.createIntent(v.getContext(), false, classId, NEGATIVE_TASK_ID)));
        doneButton.setOnClickListener(v -> finish());
    }

    @Override
    protected void onResume() {
        super.onResume();
        refreshTest(false);
        refreshHomework(false);
    }

    @Override
    public void refreshTest(boolean isDelete) {
        if (isDelete)
            StudentifyDatabaseProvider.getStudentClassDao(this).updateStudentClassTotalTest(classId, -1);

        List<Task> taskWithType = StudentifyDatabaseProvider
                .getTaskDao(this)
                .getTaskWithType(classId, TaskType.TEST.name());

        noTest.setVisibility(taskWithType.isEmpty() ? View.VISIBLE : View.GONE);
        testAdapter.updateList(taskWithType);
    }

    @Override
    public void refreshHomework(boolean isDelete) {
        if (isDelete)
            StudentifyDatabaseProvider.getStudentClassDao(this).updateStudentClassTotalHomework(classId, -1);

        List<Task> taskWithType = StudentifyDatabaseProvider
                .getTaskDao(this)
                .getTaskWithType(classId, TaskType.HOMEWORK.name());

        noHomework.setVisibility(taskWithType.isEmpty() ? View.VISIBLE : View.GONE);
        homeworkAdapter.updateList(taskWithType);
    }

    private void setupTestRV() {
        testAdapter = new TaskAdapter(this, false);
        testRecyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        testRecyclerView.setAdapter(testAdapter);
    }

    private void setupHomeworkRV() {
        homeworkAdapter = new TaskAdapter(this, true);
        homeworkRecyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        homeworkRecyclerView.setAdapter(homeworkAdapter);
    }
}

File: app/src/main/java/cheema/hardeep/sahibdeep/studentify/activities/TasksActivity.kt
package cheema.hardeep.sahibdeep.studentify.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import cheema.hardeep.sahibdeep.studentify.R
import cheema.hardeep.sahibdeep.studentify.adapters.TaskAdapter
import cheema.hardeep.sahibdeep.studentify.database.StudentifyDatabaseProvider
import cheema.hardeep.sahibdeep.studentify.interfaces.ViewRefreshInterface
import cheema.hardeep.sahibdeep.studentify.models.tables.TaskType

private const val KEY_CLASS_ID = "key-class-id"

fun createTaskActivityIntent(context: Context, classId: Int) =
        Intent(context, TasksActivity::class.java).apply {putExtra(KEY_CLASS_ID, classId)}

class TasksActivity : AppCompatActivity(), ViewRefreshInterface {

    @BindView(R.id.addHomeworkButton)
    lateinit var addHomeworkButton: ImageView

    @BindView(R.id.addTestButton)
    lateinit var addTestButton: ImageView

    @BindView(R.id.homeworkRecyclerView)
    lateinit var homeworkRecyclerView: RecyclerView

    @BindView(R.id.noTest)
    lateinit var noTest: TextView

    @BindView(R.id.noHomework)
    lateinit var noHomework: TextView

    @BindView(R.id.testRecyclerView)
    lateinit var testRecyclerView: RecyclerView

    @BindView(R.id.doneButton)
    lateinit var doneButton: Button

    lateinit var homeworkAdapter: TaskAdapter
    lateinit var testAdapter: TaskAdapter
    var classId = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_tasks)
        supportActionBar?.hide()
        ButterKnife.bind(this)
        classId = intent.getIntExtra(KEY_CLASS_ID, -1)

        setupHomeworkRV()
        setupTestRV()

        addHomeworkButton.setOnClickListener { v: View -> startActivity(TasksDetailsActivity.createIntent(v.context, true, classId, TasksDetailsActivity.NEGATIVE_TASK_ID)) }
        addTestButton.setOnClickListener { v: View -> startActivity(TasksDetailsActivity.createIntent(v.context, false, classId, TasksDetailsActivity.NEGATIVE_TASK_ID)) }
        doneButton.setOnClickListener { v: View? -> finish() }
    }

    override fun onResume() {
        super.onResume()
        refreshTest(false)
        refreshHomework(false)
    }

    override fun refreshTest(isDelete: Boolean) {
        if (isDelete) StudentifyDatabaseProvider.getStudentClassDao(this).updateStudentClassTotalTest(classId, -1)
        val taskWithType = StudentifyDatabaseProvider
                .getTaskDao(this)
                .getTaskWithType(classId, TaskType.TEST.name)
        noTest.visibility = if (taskWithType.isEmpty()) View.VISIBLE else View.GONE
        testAdapter.updateList(taskWithType)
    }

    override fun refreshHomework(isDelete: Boolean) {
        if (isDelete) StudentifyDatabaseProvider.getStudentClassDao(this).updateStudentClassTotalHomework(classId, -1)
        val taskWithType = StudentifyDatabaseProvider
                .getTaskDao(this)
                .getTaskWithType(classId, TaskType.HOMEWORK.name)
        noHomework.visibility = if (taskWithType.isEmpty()) View.VISIBLE else View.GONE
        homeworkAdapter.updateList(taskWithType)
    }

    private fun setupTestRV() {
        testRecyclerView.apply {
            layoutManager = LinearLayoutManager(this@TasksActivity, RecyclerView.VERTICAL, false)
            adapter = TaskAdapter(this@TasksActivity, false).also { testAdapter = it }
        }
    }

    private fun setupHomeworkRV() {
        homeworkRecyclerView.apply {
            layoutManager = LinearLayoutManager(this@TasksActivity, RecyclerView.VERTICAL, false)
            adapter = TaskAdapter(this@TasksActivity, true).also { homeworkAdapter = it }
        }
    }



}
